packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -4337,7 +4347,21 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
     // notifications.
     _flushRouteAnnouncement();
 
+    final _RouteEntry? lastEntry = _lastRouteEntryWhereOrNull(_RouteEntry.isPresentPredicate);
+    if (lastEntry != null && _lastEntry != lastEntry) {
+      for (final NavigatorObserver observer in _effectiveObservers) {
+        observer.didChangeTop(lastEntry.route, _lastEntry?.route);
+      }
+    }
+    _lastEntry = lastEntry;
     // Announce route name changes.
+    if (widget.reportsRouteUpdateToEngine) {
+      final String? routeName = lastEntry?.route.settings.name;
+      if (routeName != null && routeName != _lastAnnouncedRouteName) {
+        SystemNavigator.routeInformationUpdated(uri: Uri.parse(routeName));
+        _lastAnnouncedRouteName = routeName;
+      }
+    }
```

Comment:
Should this just be part of the `if` below?

Reply:
Also, why are we calling `SystemNavigator.routeInformationUpdated` twice, in this `if` and again below in the other if? Should this `if` replace the `if` below?